File: skellig-feature/src/main/java/org/skellig/feature/parser/DefaultTestPreRequisitesExtractor.java
package org.skellig.feature.parser;

import org.skellig.feature.TagDetails;
import org.skellig.feature.TestPreRequisites;

import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class DefaultTestPreRequisitesExtractor {

    private Collection<TestPreRequisitesExtractor<? extends TestPreRequisites<?>>> testPreRequisitesExtractors;
    private TagDetailsExtractor tagDetailsExtractor = new TagDetailsExtractor();

    public DefaultTestPreRequisitesExtractor() {
        testPreRequisitesExtractors =
                Stream.of(
                        tagDetailsExtractor,
                        new InitDetailsExtractor(),
                        new DataDetailsExtractor()
                ).collect(Collectors.toList());
    }

    public List<TestPreRequisites<?>> extractFrom(String text) {
        List<TestPreRequisites<?>> preRequisites =
                testPreRequisitesExtractors.stream()
                        .map(extractor -> extractor.extractFrom(text))
                        .filter(Objects::nonNull)
                        .collect(Collectors.toList());
        return preRequisites.isEmpty() ? null : preRequisites;
    }

    public Set<String> extractTags(String text) {
        final TestPreRequisites<TagDetails> tags = tagDetailsExtractor.extractFrom(text);
        return tags != null ? tags.getDetails().getTags() : null;
    }
}

File: skellig-feature/src/main/kotlin/org/skellig/feature/parser/DefaultTestPreRequisitesExtractor.kt
package org.skellig.feature.parser

import org.skellig.feature.TestPreRequisites

class DefaultTestPreRequisitesExtractor {

    private var testPreRequisitesExtractors: Collection<TestPreRequisitesExtractor<out TestPreRequisites<*>>>? = null
    private val tagDetailsExtractor = TagDetailsExtractor()

    init {
        testPreRequisitesExtractors = listOf(tagDetailsExtractor)
    }

    fun extractFrom(text: String?): List<TestPreRequisites<*>>? {
        val preRequisites: List<TestPreRequisites<*>> = testPreRequisitesExtractors!!
                .mapNotNull { it.extractFrom(text) }
                .toList()
        return if (preRequisites.isEmpty()) null else preRequisites
    }

    fun extractTags(text: String?): Set<String>? {
        val tags = tagDetailsExtractor.extractFrom(text)
        return tags?.getDetails()?.tags
    }
}